docs/2024-03-14-shardy-partitioner-rfc.md

```diff
@@ -0,0 +1,358 @@
+
+Let’s explore the sharding representation with a simple rank 2 tensor and 4 devices.
+
+We first reshape the 4 devices `[0, 1, 2, 3] `into a 2-d array `[[0, 1], [2, 3]]` to create a mesh with 2 axes:
+
+```c++
+@mesh_xy = <"x"=2, "y"=2>
+```
+
+We can then shard the following rank 2 tensor `[[a, b], [c, d]]` as follows:
+
+![Figure1](images/2024-03-14-shardy-partitioner-rfc/sdy0.png)
+
+#### Other key components
+
+* **Open/Closed dimensions** - dimensions can either be open - can be further sharded on available axes; or closed - are fixed and can’t be changed.
+* **Explicitly replicated axes** - all axes that are not used to shard a dimension are implicitly replicated, but the sharding can specify axes that are explicitly replicated and therefore cannot be used to shard a dimension later on.
+* **Axis splitting and sub-axes** - a (full) mesh axis can be split into multiple sub-axes that can be individually used to shard a dimension or be explicitly replicated.
+* **Multiple logical meshes** - different shardings can be bound to different logical meshes, which can have different axes or even a different order of logical device ids.
+* **Priorities** - to partition a program incrementally, priorities can be attached to dimension shardings, which determine in which order per-dimension sharding constraints will be propagated throughout the module.
+* **Dimension sharding divisibility** - a dimension can be sharded on axes whose product of sizes doesn’t divide the dimension size.
+
+
+## Detailed Design
+
+We expand the basic structure and each key component in this section.
+
+
+### Basic structure
+
+The dimension shardings tell us for each dimension of the tensor, along which axes (or"sub-axes") it is sharded from major to minor. All other axes that don’t shard a dimension are implicitly replicated (or explicitly replicated)
+
+We will start with a simple example and extend it as we describe additional features.
+
+```c++
+@mesh_xy= <"x"=2, "y"=4, "z"=2>
+
+// The 1st tensor dimension is sharded along axis "x" and the 2nd tensor dimension is
+// sharded along axis "z" then further along axis "y". The local shape of this tensor (i.e. the shape on a single device), would be tensor<2x1xf32>.
+sharding<@mesh_xy, [{"x"}, {"z", "y"}]> : tensor<4x8xf32>
+```
+
+#### Invariants
+
+*   The number of dimension shardings must match the rank of the tensor.
+*   All axis names must exist in the referenced mesh.
+*   Axes or sub-axes can only appear once in the sharding representation (each one either shards a dimension or is "explicitly replicated".
+
+
+### Open/closed dimensions
+
+Each dimension of a tensor can either be open or closed.
+
+
+#### Open
+
+An open dimension is one open for propagation to further shard it along additional axes, i.e. the specified dimension sharding doesn’t have to be the final sharding of that dimension. This is similar (but not exactly the same as) to
+
+*   GSPMD’s <code>unspecified_dims</code>
```

Comment:
Instead of the `<code>` syntax could we use back-ticks \`\`? Here and throughout